File: python/taflow/metrics/longest_losing_streak.py
"""Longest strictly losing observation run."""
from __future__ import annotations
from typing import Any
from .._native.metrics import LongestLosingStreak as _Native
from ._input import as_metric_series


class LongestLosingStreak:
    """Compute the longest consecutive run of negative observations.

    The independent oracle is QuantStats 0.0.81 ``consecutive_losses``.
    Strict negatives extend a streak; zero, positive values, and breakevens
    break it. Warm-up requires one usable observation, after which a no-loss
    sample returns integer zero. Inputs are explicit period returns, raw period
    P&L, or closed-trade P&L. Rust owns O(1) state; no annualization occurs.
    """
    def __init__(self, nan_policy: str = "omit") -> None:
        """Initialize an empty configured metric."""
        self._state = _Native(nan_policy)

    def from_returns(self, returns: Any, *, column: str | None = None) -> "LongestLosingStreak":
        """Append chronological returns observations and return this metric."""
        self._state.from_returns(as_metric_series(returns, column=column))
        return self

    def from_pnl(self, pnl: Any, *, column: str | None = None) -> "LongestLosingStreak":
        """Append chronological pnl observations and return this metric."""
        self._state.from_pnl(as_metric_series(pnl, column=column))
        return self

    def from_trades(self, trades: Any, *, column: str | None = None) -> "LongestLosingStreak":
        """Append chronological trades observations and return this metric."""
        self._state.from_trades(as_metric_series(trades, column=column))
        return self

    def append(self,value:float)->"LongestLosingStreak":
        """Append one observation and return this metric."""
        self._state.append(float(value));return self
    def extend(self,values:Any,*,column:str|None=None)->"LongestLosingStreak":
        """Append observations and return this metric."""
        self._state.extend(as_metric_series(values,column=column));return self
    @property
    def value(self)->int|None:
        """Return the longest run, or ``None`` during warm-up."""
        return self._state.value
    def compute(self)->int|None:
        """Return current scalar without replaying input."""
        return self._state.compute()
    def reset(self)->"LongestLosingStreak":
        """Clear state and return this metric."""
        self._state.reset();return self
    def __len__(self)->int:
        """Return valid observation count delegated to Rust."""
        return len(self._state)


__all__=["LongestLosingStreak"]

File: crates/taflow-metrics/src/metrics/longest_losing_streak.rs
use crate::{MetricError, MetricInputKind, MetricInputState, MetricResult, NanPolicy};

/// Longest run of strictly negative observations.
#[derive(Debug, Clone)]
pub struct LongestLosingStreak {
    input: MetricInputState,
    current: usize,
    longest: usize,
}
impl LongestLosingStreak {
    /// Construct an empty state for returns, raw period P&L, or closed trades.
    pub fn new(nan_policy: NanPolicy) -> MetricResult<Self> {
        Ok(Self {
            input: MetricInputState::unbound(nan_policy),
            current: 0,
            longest: 0,
        })
    }
    /// Append one observation and return the longest losing run.
    pub fn append(&mut self, value: f64) -> MetricResult<Option<usize>> {
        if let Some(observation) = self.input.append(value)? {
            if observation < 0.0 {
                self.current += 1;
                self.longest = self.longest.max(self.current);
            } else {
                self.current = 0;
            }
        }
        Ok(self.value())
    }
    /// Append observations through the same state.
    pub fn extend(&mut self, values: &[f64]) -> MetricResult<Option<usize>> {
        for &value in values {
            self.append(value)?;
        }
        Ok(self.value())
    }
    /// Return longest losing run; a nonempty no-loss sample returns zero.
    pub fn value(&self) -> Option<usize> {
        (!self.input.is_empty()).then_some(self.longest)
    }
    /// Return current scalar without replay.
    pub fn compute(&self) -> Option<usize> {
        self.value()
    }
    /// Reset observations and streaks.
    pub fn reset(&mut self) {
        self.input.reset();
        self.current = 0;
        self.longest = 0;
    }
    /// Return valid observation count.
    pub fn len(&self) -> usize {
        self.input.len()
    }
    /// Return whether empty.
    pub fn is_empty(&self) -> bool {
        self.input.is_empty()
    }
}

crate::impl_observation_metric_lifecycle!(LongestLosingStreak);
